feature: produce and consume fcm token registered event

eTaphee opened the pull request:
<!-- 필요한 경우 팀원들과 의논하여 양식을 변경할 수 있습니다. -->
<!-- PR은 상세히 적어주시는게 좋습니다. -->
<!-- 이해를 돕기위한 이미지를 써도 좋습니다. -->

## Changes
<!-- 어떤점이 변경되었는지 적어주세요. -->
- Producer(`API`)
   - POST /api/notifications/tokens을 통해 토큰을 등록하게 되면, 토큰 생성 이벤트 메시지를 발행합니다.
- Consumer(`FcmTokenRegisteredEventListener`)
  - 이벤트로 넘겨온 토큰이 DB에 있는지 조회하고 없으면, 예외를 발생하고 해당 메시지가 재처리 되지 않도록 `CustomErrorHandler`를 등록했습니다. 
  - 사용자가 팔로우한 보호소 목록을 가져와 구독할 토픽이름을 생성합니다. ex) topic.shelters.{id}
  - 사용자가 북마크한 게시글 목록을 가져와 구독할 토픽이름을 생성합니다. ex) topic.adoptions.{id}
  - firebase sdk 를 이용하여 생성된 토픽 목록을 구독합니다.
  - 구독 처리 결과를 FcmSubscription 엔티티로 저장합니다.

## Background
<!-- 이 PR이 진행된 배경을 적어주세요. -->
사용자가 새로운 기기에서 로그인하거나, 로그인된 기기에서 fcm 토큰이 갱신되는 경우
해당 기기에서도 본인이 설정한 푸시 알림을 받을 수 있도록, 토픽 구독을 처리해야합니다.

## Discuss
<!-- 토론할 내용이 있다면 적어주세요. -->
Firebase SDK를 사용하기 위해 인증정보가 필요합니다. 인증 정보는 `GOOGLE_APPLICATION_CREDENTIALS` 환경 변수에 제가 슬랙에 올린 json 파일의 경로를 넣어주시면 됩니다.
<img width="661" alt="image" src="https://github.com/banzzokee/banzzokee-back/assets/43163409/13774311-4bdf-47af-999b-4edbd17a5bf2">

## Issues
<!-- 관련된 이슈를 #{issueNumber}를 통해 태그해주세요. -->
<!-- ex) 해결한 이슈: #1, #2 -->
<!-- ex) 구현이 필요한 이슈: #3, #4 -->
해결한 이슈: #82
구현이 필요한 이슈: #81

## Execute
<!-- 실행된 결과에 대해 적어주세요. -->
<!-- 어떻게 테스트를 진행했는지, 어떤 검토를 거쳤는지 적어주시면 좋습니다. -->
<!-- 테스트 코드는 중요합니다! -->
테스트
- 토큰 등록 시 이벤트 발생 테스트
  - 토큰 등록 시  `ApplicationEventPublisher`를 통해 `FcmTokenRegisteredEvent`가 호출되는지 검증합니다.
- 토큰 등록 이벤트 처리 테스트
  - 이벤트로 넘겨 받은 payload의 토큰을 db에서 조회했을 때 없으면 예외를 발생합니다.
  - 정상적으로 처리됐을 경우 팔로우하는 보호소 개수만큼 `FcmService.subscribeToTopic`가 호출되는지 검증합니다.
  - repository에 save되는 구독 entity 정보를 captor를 통해 검사합니다.
 
<img width="1073" alt="image" src="https://github.com/banzzokee/banzzokee-back/assets/43163409/efbe1a34-c804-42e0-87a1-c74f6243a4a6">

<img width="1024" alt="image" src="https://github.com/banzzokee/banzzokee-back/assets/43163409/8188c557-2600-44ad-82da-f11abdc24693">

## Reference
<!-- 참조한 레퍼런스가 있다면 적어주세요. -->
AMQP 메시지 수신 예외 처리 전략
https://www.baeldung.com/spring-amqp-error-handling

eTaphee committed:
feature: fcm 토큰 produce, consume 구현

FCM 토큰 등록 시, 보호소 목록 및 북마크 목록에 대해 토픽 구독하고 구독 정보를 저장한다.

eTaphee committed:
test: fcm 토큰 등록 이벤트 처리 테스트 작성

eTaphee committed:
feature: unsubscribeFromTopic 메서드 추가

nadoran78 reviewed (approved):
알림이 제법 복잡한 기능이었군요.. 고생 많으셨습니다!

eTaphee committed:
chore: 토픽 관리 큐 이름 변경, 바인딩 룰 추가

- queue.manage.fcm.subscription -> queue.manage.fcm.topic 변경
- 바인딩 룰이 누락되어 추가

eTaphee committed:
refactor: BaseMessage 리팩터링

- 라우팅 키를 이벤트 핸들러에서 명시해서 보내는 방법에서 이벤트 클래스에서 getRoutingKey() 를 오버라이딩 하는것으로 변경
- 메시지 포맷이 같으므로, 각각의 메시지를 처리하는 핸들러 대신 BaseMessage를 처리하는 MessageHandler 추가
- MessageHandler 추가로 FcmTokenRegisteredEventHandler 삭제

eTaphee committed:
feature: 보호소 사용자 팔로우, 언팔로우  produce, consume 구현

- 팔로우, 언팔로우 행위는 결국 topic를 subscribe/unsubscribe하는 행위이므로 FcmTopicStatusChangeEvent 클래스로 추상화
- FcmTopicStatusChangeEvent 클래스를 상속받아 상세 이벤트 클래스 구현
- FcmTopicStatusChangeEvent 이벤트 처리 핸들러 구현
- FcmTokenRegisteredEventListener 클래스도 topic를 subscribe/unsubscribe하기 떄문에 같은 Consumer 클래스로 묶음

eTaphee committed:
refactor: 패키지 명 수정 누락 반영

eTaphee committed:
test: 사용자 팔로우 테스트 수정 및 언팔로우 테스트 케이스 작성

- 팔로우 시 발행되는 ShelterUserFollowedEvent 이벤트 검증
- 언팔로우 이벤트 추가
  - 팔로우 정보가 없으면 delete, publishEvent 호출하지 않음
  - 팔로우 정보가 있으면 delete, publishEvent 호출

eTaphee committed:
test: 토픽 상태 변경 처리 테스트 작성

- 액션이 SUBSCRIBE 시 subscribeTopicsAndSave 검증
- 액션이 UNSUBSCRIBE 시 unsubscribeTopicsAndDelete 검증

eTaphee committed:
fix: 자잘한 버그 및 기타 수정

- follower, followee 헷갈려서 바꿔쓴 코드 수정
- topic 필드가 있고 getTopic()이 이름이 겹처서 역직렬화 안되는 문제가 있어서 topicCategory로 변경

eTaphee committed:
refactor: event 클래스 단순화

불필요한 상속 클래스 제거

jnjeaaaat commented:
수고 많으셨습니다~!

eTaphee committed:
Merge pull request #87 from banzzokee/feature/produce-and-consume-shelter-followed-event

feature: produce and consume shelter followed event
